contracts/extensions/MultisigPermissions.sol

```diff
@@ -0,0 +1,541 @@
+      if (motion.domainSkillId == 0 && motion.requiredPermissions == 0){
+        motion.domainSkillId = actionDomainSkillId;
+        motion.requiredPermissions = actionRequiredPermissions;
+      } else {
+        require(motion.domainSkillId == actionDomainSkillId && motion.requiredPermissions == actionRequiredPermissions, "colony-multisig-invalid-motion");
+      }
+    }
+
+    // Get permissions in _permissionDomainId for user
+    bytes32 userPermissions = getUserRoles(msgSender(), _permissionDomainId);
+    // Work out which domain they're using their permissions in, via child skill index
+    uint256 permissionSkillId = colony.getDomain(_permissionDomainId).skillId;
+    uint256 userActingInDomainWithSkillId = colonyNetwork.getChildSkillId(permissionSkillId, _childSkillIndex);
+
+    // Check it's same as the domain being acted on by _motionId
+    require(userActingInDomainWithSkillId == motion.domainSkillId, "colony-multisig-not-same-domain");
+
+    // Check they have the permissions
+    require((userPermissions & motion.requiredPermissions) != bytes32(0), "colony-multisig-no-permissions");
+
+    emit MotionCreated(msgSender(), motionCount);
+
+    // TODO: do this per-permission
+    changeApproval(_permissionDomainId, _childSkillIndex, motionCount, true);
+  }
+
+  function changeApproval(uint256 _permissionDomainId, uint256 _childSkillIndex, uint256 _motionId, bool _approved) public notExecuted(_motionId) {
+    // Get permissions in _permissionDomainId for user
+    bytes32 userPermissions = getUserRoles(msgSender(), _permissionDomainId);
+
+    // Work out which domain they're using their permissions in, via child skill index
+    uint256 permissionSkillId = colony.getDomain(_permissionDomainId).skillId;
+    uint256 userActingInDomainWithSkillId = colonyNetwork.getChildSkillId(permissionSkillId, _childSkillIndex);
+
+    // Check it's same as the domain being acted on by _motionId
+    Motion storage motion = motions[_motionId];
+    require(userActingInDomainWithSkillId == motion.domainSkillId, "colony-multisig-not-same-domain");
+
+    // If they're trying to approve, needs some overlap of permissions
+    if (_approved){
+      require(userPermissions & motion.requiredPermissions > 0, "colony-multisig-no-permissions");
+    }
+
+    uint8 roleIndex = 0;
+    bool newlyApproved = false;
+    bool anyBelowThreshold = false;
+    while (uint256(motion.requiredPermissions) >= (1 << roleIndex)){
+      if ((motion.requiredPermissions & bytes32(1<<roleIndex)) > 0){
+        // Then the motion requires this permission. Let's check it
+
+        // The user either needs to have the permission, or are removing an approval
+        // If not, move on to next permissions
+        if (!(uint256(userPermissions) & (1 << roleIndex) != 0 || !_approved)) {
+          roleIndex += 1;
+          continue;
+        }
+
+        // Update appropriately
+        if (motionApprovals[_motionId][msgSender()][ColonyRole(roleIndex)] != _approved){
```

Comment:
Wondering if we could store all approvals in a single bytes32, rather than many different booleans. Seems like it would be cheaper, gas-wise, when a user might approve with many permissions.

Reply:
Yeah, this seems like a great idea :+1: